felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -1567,3 +1721,110 @@ func (m *bpfEndpointManager) getInterfaceIP(ifaceName string) (*net.IP, error) {
 	}
 	return nil, errors.New("interface ip address not found")
 }
+
+func (m *bpfEndpointManager) onServiceUpdate(update *proto.ServiceUpdate) {
+	log.WithFields(log.Fields{
+		"Name":      update.Name,
+		"Namespace": update.Namespace,
+	}).Info("Service Update")
+
+	ips := make([]string, 0, 2+len(update.ExternalIps))
+	if update.ClusterIp != "" {
+		ips = append(ips, update.ClusterIp)
+	}
+	if update.LoadbalancerIp != "" {
+		ips = append(ips, update.LoadbalancerIp)
+	}
+	ips = append(ips, update.ExternalIps...)
+
+	key := serviceKey{name: update.Name, namespace: update.Namespace}
+	m.dirtyServices[key] = ips
+}
+
+func (m *bpfEndpointManager) onServiceRemove(update *proto.ServiceRemove) {
+	log.WithFields(log.Fields{
+		"Name":      update.Name,
+		"Namespace": update.Namespace,
+	}).Info("Service Remove")
+
+	key := serviceKey{name: update.Name, namespace: update.Namespace}
+	m.dirtyServices[key] = nil
+}
+
+func (m *bpfEndpointManager) reconcileServices() {
+	for svc, ips := range m.dirtyServices {
+		errored := false
+		for _, ip := range ips {
+			if err := m.dp.setRoute(ip); err != nil {
+				log.WithError(err).Errorf("Failed to set route to %s via bpfnatin.", ip)
+				errored = true
+			}
+		}
+
+		known := m.services[svc]
+
+		for _, old := range known {
```

Comment:
Wonder if we can see services with overlapping IPs (especially ext IPs) for some transient window.  I suspect we can't rule it out, in which case I think we need to do something to handle it; otherwise I think you'll end up with no route after one of the overlapping service is removed